connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java

```
@@ -429,7 +429,9 @@ void enrich(ConfigDef newDef) {
                 final ConfigDef.Validator typeValidator = ConfigDef.LambdaValidator.with(
                     (String name, Object value) -> {
                         validateProps(prefix);
-                        getConfigDefFromConfigProvidingClass(typeConfig, (Class<?>) value);
+                        if (value != null) {
```

The value here is null if the class couldn't be loaded, in which case, it's not necessary to try to do any further kind of validation.

---

Ah, this hides the exception message from "Not a (something)" and only shows the "Missing required configuration (name) which has no default value". I think that is reasonable.

The other call-site for this method is ConnectorConfig.EnrichablePlugin which swallows this error on the validate() code path and propagates them when instantiating the ConnectorConfig. Instantiating the ConnectorConfig will also throw the "Missing required configuration" error, so it is not necessary to throw the error.

I think you could safely change the getConfigDefFromConfigProvidingClass implementation to return an empty stream when the value is null, rather than throwing an exception. I don't think this is necessary, but maybe it keeps these two code paths more similar.

---

> I think you could safely change the getConfigDefFromConfigProvidingClass implementation to return an empty stream when the value is null, rather than throwing an exception.

Are you sure? It looks like the error swallowing in `ConfigDef.EnrichablePlugin::populateConfigDef` takes place conditionally, and we still do throw exceptions that originate from the `getConfigDefFromConfigProvidingClass` sometimes.

---

> It looks like the error swallowing in ConfigDef.EnrichablePlugin::populateConfigDef takes place conditionally

Yeah, the condition depends on whether validation is being executed or whether the full ConnectorConfig is being constructed.

> we still do throw exceptions that originate from the getConfigDefFromConfigProvidingClass sometimes.

I think the other exceptions for non-null classes are fine to leave as-is. I think we could only reasonably change the null-class behavior.

---

Ah sorry, I misread your comment!

I'm hesitant to change the behavior of instantiating a `ConnectorConfig` unless it provides a practical, positive impact for users. It doesn't seem significantly cleaner to move the null check either, since guarding against null `Class<?>` instances is an inherent necessity of the current `ConfigDef.Validator` API (and one that's taken into account with validator classes added in https://github.com/apache/kafka/pull/14304).

I've added a comment indicating when we expect `null` values in the validator. I'm hopeful that this should be enough to avoid adding to the cognitive load with this change.